File: leetcode3/최원준/463. Island Perimeter.py
class Solution:
    def islandPerimeter(self, grid: list[list[int]]) -> int:
        n = len(grid)
        m = len(grid[0])
        dx = [0,0,1,-1]
        dy = [1,-1,0,0]

        def bfs(row, col):
            ans = 0
            grid[row][col] = 2
            queue = Deque()
            queue.append([row,col])

            while queue:
                x, y = queue.popleft()
                perimeter = 4
                for dirs in range(4):
                    nx = x + dx[dirs]
                    ny = y + dy[dirs]

                    if nx < 0 or n <= nx or ny < 0 or m <= ny:
                        continue
                    elif 0<=nx<n and 0<=ny<m:
                        if grid[nx][ny] == 1:
                            queue.append([nx,ny])
                            grid[nx][ny] = 2
                            perimeter-=1
                        elif grid[nx][ny] == 2:
                            perimeter-=1
                ans += perimeter
            return ans
                        

        
        for i in range(n):
            for j in range(m):
              if grid[i][j] == 1:
                return bfs(i, j)  

        return 0

File: leetcode3/황은지/463. Island Perimeter.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var islandPerimeter = function(grid) {

    const ROW=grid.length;
    const COL=grid[0].length;

    //BFS
    function doBFS(startR,startC){
        const DIR=[[-1,0],[0,1],[1,0],[0,-1]]
        const visited=Array.from({length:ROW},()=>Array(COL));
        let count=0;

        visited[startR][startC]=true;
        
        const queue=[[startR,startC]];
        let head=0;

        while(head<queue.length){
            const [curR,curC]=queue[head++];

            for(let i=0;i<4;i++){
                const [nextR,nextC]=[curR+DIR[i][0],curC+DIR[i][1]];
                if(nextR<0 || nextR>=ROW || nextC<0 || nextC>=COL
                || grid[nextR][nextC]===0){
                    count++;
                    continue;
                }
                if(visited[nextR][nextC]) continue;
                visited[nextR][nextC]=true;
                queue.push([nextR,nextC]);
            }
        }
        return count;

    }


    // 실행부
    for(let i=0;i<ROW;i++){
        for(let j=0;j<COL;j++){
            if(grid[i][j]===1) return doBFS(i,j);
        }
    }

    
};